### PR TITLE
.github/workflows: Create workflow to CI kernel builds

### PR DESCRIPTION
**Pull request by 6by9**
Builds the bcmrpi, bcm2709, bcm2711, and bcm2835 32 bit kernels, and defconfig and bcm2711 64bit kernels, saving the artifacts for 7 days.

### CODE REVIEW

**Comment by 6by9** [comment]
I'm just waiting on https://github.com/6by9/linux/actions/runs/3462901730 to complete to ensure I haven't made any last minute typos.

This will need workflows enabling on the repo, and then any PR should trigger an automatic build of the 6 main kernel variants. I haven't found a way to check for any new warnings though, but there must be one.

Adding any extra jobs is relatively easy, so adding a pelwell-bot for overlays needing a README update should be possible :-)

**Comment by 6by9** [comment]
Hold fire - looking at adding an exclude to the .github directory so that changing the workflow script does not trigger the action.

File: .github/workflows/kernel-build.yml [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+        fetch-depth: 1
+        clean: true
+
+    - name: Build kernel
+      run: |
+        mkdir ${{github.workspace}}/build
+        make ARCH=arm64 KERNEL=kernel8 CROSS_COMPILE=aarch64-linux-gnu- O=${{github.workspace}}/build defconfig
+        make ARCH=arm64 KERNEL=kernel8 CROSS_COMPILE=aarch64-linux-gnu- O=${{github.workspace}}/build -j ${{env.NUM_JOBS}} Image.gz modules dtbs
+        mkdir -p ${{github.workspace}}/install/boot
+        make ARCH=arm KERNEL=kernel8 CROSS_COMPILE=aarch64-linux-gnu- O=${{github.workspace}}/build INSTALL_MOD_PATH=${{github.workspace}}/install modules_install
```
Comment (nbuchwitz) [inline .github/workflows/kernel-build.yml]: ARCH=arm => ARCH=arm64

File: .github/workflows/kernel-build.yml [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+        fetch-depth: 1
+        clean: true
+
+    - name: Build kernel
+      run: |
+        mkdir ${{github.workspace}}/build
+        make ARCH=arm64 KERNEL=kernel8 CROSS_COMPILE=aarch64-linux-gnu- O=${{github.workspace}}/build bcm2711_defconfig
+        make ARCH=arm64 KERNEL=kernel8 CROSS_COMPILE=aarch64-linux-gnu- O=${{github.workspace}}/build -j ${{env.NUM_JOBS}} Image.gz modules dtbs
+        mkdir -p ${{github.workspace}}/install/boot
+        make ARCH=arm KERNEL=kernel8 CROSS_COMPILE=aarch64-linux-gnu- O=${{github.workspace}}/build INSTALL_MOD_PATH=${{github.workspace}}/install modules_install
```
Comment (nbuchwitz) [inline .github/workflows/kernel-build.yml]: ARCH=arm => ARCH=arm64

**Comment by nbuchwitz** [comment]
Nice work! Out of experience from our RevPi builds, I would suggest to refactor the workflows with matrix builds. This will reduce duplication to a minimum and improve maintainability. arch and defconfig would be the minimal dimensions (https://docs.github.com/en/actions/using-jobs/using-a-matrix-for-your-jobs)

**Comment by 6by9** [comment]
Updated to correct the ARM/ARM64 errors, and also adds the dtoverlay check status badges to the base README.

I've looked at refactoring for a matrix, but haven't got the magic working to switch the KERNEL= vars, nor anything sensible for the mainline 64bit build that wants defconfig rather than <thing>_defconfig.

File: .github/workflows/kernel-build.yml [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+        fetch-depth: 1
+        clean: true
+
+    - name: Build kernel
+      run: |
+        mkdir ${{github.workspace}}/build
+        make ARCH=arm KERNEL=kernel7l CROSS_COMPILE=arm-linux-gnueabihf- O=${{github.workspace}}/build bcm2711_defconfig
+        make ARCH=arm KERNEL=kernel7l CROSS_COMPILE=arm-linux-gnueabihf- O=${{github.workspace}}/build -j ${{env.NUM_JOBS}} zImage modules dtbs
+        mkdir -p ${{github.workspace}}/install/boot
+        make ARCH=arm64 KERNEL=kernel7l CROSS_COMPILE=arm-linux-gnueabihf- O=${{github.workspace}}/build INSTALL_MOD_PATH=${{github.workspace}}/install modules_install
```
Comment (pelwell) [inline .github/workflows/kernel-build.yml]: `ARCH=arm`?

File: .github/workflows/kernel-build.yml (hunk 1 above)
Comment (pelwell) [reply .github/workflows/kernel-build.yml]: Still.

Comment (6by9) [reply .github/workflows/kernel-build.yml]: I went and fixed the wrong instance :-(
Should be sorted now.

File: .github/workflows/kernel-build.yml (hunk 3 above)
Comment (6by9) [reply .github/workflows/kernel-build.yml]: Fixed the wrong line :-(

**Comment by nbuchwitz** [comment]
> Updated to correct the ARM/ARM64 errors, and also adds the dtoverlay check status badges to the base README.
> 
> I've looked at refactoring for a matrix, but haven't got the magic working to switch the KERNEL= vars, nor anything sensible for the mainline 64bit build that wants defconfig rather than _defconfig.

@6by9 I was thinking sth like this: https://github.com/nbuchwitz/actions-examples/blob/main/.github/workflows/test1.yml

**Pull request merged**